MXD Backend Service API

ravinderkumarsap opened the pull request:
## Description 
Please describe your PR:  
Created a http backend service to validate the transfer.
Link for issue  : https://github.com/eclipse-tractusx/tutorial-resources/issues/128

- What does this PR introduce?  

 Here's a step-by-step Explanation of created a REST API application to complete and validate a successful asset transfer:

Create a Content Storage API:

POST API /api/v1/contents:
{{backend-service-url}}/api/v1/content/

Endpoint to store assets.
Accepts asset data in the request body.
Persists the asset data and returns a URL or ID in the response.

GET API /api/v1/contents/{id}:
{{backend-service-url}}/api/v1/content/1

Endpoint to fetch content by ID.
Takes the content ID as a path parameter.
Returns the stored content associated with the given ID.
Create a Transfer API:

POST API /api/v1/transfers:
{{backend-service-url}}/api/v1/transfer/
Endpoint to accept transfer data from the connector.
Expects a JSON payload similar to:
{
  "id": "<transfer id>",
  "endpoint": "http://alice-tractusx-connector-dataplane:8081/api/public",
  "authKey": "Authorization",
  "authCode": "<Auth Code>",
  "properties": {}
}

Persists the transfer data with the transfer ID.
GET API /api/v1/transfers/{id}:
{{backend-service-url}}/api/v1/transfer/52ba4b2f-ae3a-46ae-96f2-6781aa819386/

Endpoint to retrieve transfer data by ID.
Takes the transfer ID as a path parameter.
Returns the JSON data pushed by the connector for the specified transfer.

GET API /api/v1/transfers/{id}/contents:
{{backend-service-url}}/api/v1/transfer/52ba4b2f-ae3a-46ae-96f2-6781aa819386/contents/

Endpoint to retrieve the actual assets content.
Uses the stored endpoint URL from the transfer data to fetch content.
Returns the content fetched from the specified endpoint.

GET API /api/v1/content/random
{{backend-service-url}}/api/v1/content/random
Return the Sample Random Generated JSON 

 Does it add a new feature?   yes
 Is it enhancing documentation : yes

It has the postman collections inside postman folder inside backend-service folder. 

Closes  #128

<!-- Please tag the related issue `Fixes or Updates #issue_number`, if applicable. -->

## Pre-review checks

Please ensure to do as many of the following checks as possible, before asking for committer review:

- [x] DEPENDENCIES are up-to-date. [Dash license tool](https://github.com/eclipse/dash-licenses). Committers can open IP issues for restricted libs.
- [ ] [Copyright and license header](https://eclipse-tractusx.github.io/docs/release/trg-7/trg-7-02) are present on all affected files

hemantxpatel commented on mxd/backend-service/src/main/java/com/eclipse/mxd/repository/Impl/ContentRepositoryImpl.java:
```diff
@@ -0,0 +1,152 @@
+    private static final Logger logger = Logger.getLogger(ContentRepositoryImpl.class.getName());
+
+    @Override
+    public Long createAsset(String assetName) {
```
Rename `assetName` to `asset`.

ravinderkumarsap replied:
variable is renamed as suggested.

ravinderkumarsap commented on mxd/backend-service/Dockerfile:
```diff
@@ -0,0 +1,42 @@
+# application.properties
```
copyright header has been removed

ravinderkumarsap commented on mxd/backend-service/src/main/java/com/eclipse/mxd/BackendApplication.java:
```diff
@@ -0,0 +1,34 @@
+ * 
+ ******************************************************************************/
+
+package com.eclipse.mxd;
```
Package is stared with org.eclipse

ravinderkumarsap commented on mxd/backend-service/src/main/java/com/eclipse/mxd/Main.java:
```diff
@@ -0,0 +1,62 @@
+    private static final Logger logger = Logger.getLogger(Main.class.getName());
+
+    public static void main(String[] args) {
+        contentService.getAll();
```
getAll() method call has been removed from main method

ravinderkumarsap commented on mxd/backend-service/src/main/java/com/eclipse/mxd/Main.java:
```diff
@@ -0,0 +1,62 @@
+    }
+
+    private static void startJettyServer() {
+        int port = 8080;
```
Its coming from properties file

ravinderkumarsap commented on mxd/backend-service/src/main/java/com/eclipse/mxd/model/TransferRequest.java:
```diff
@@ -0,0 +1,92 @@
+    private String endpoint;
+    private String authKey;
+    private String authCode;
+    private PropertiesModel properties;
```
Property model is required for now

ravinderkumarsap commented on mxd/backend-service/src/main/java/com/eclipse/mxd/model/TransfersModel.java:
```diff
@@ -0,0 +1,115 @@
+
+    public TransfersModel(Long id, String asset, String contents, String transferid, Date createdDate,
+                          Date updatedDate) {
+        super();
```
extra super() call has been removed

hemantxpatel commented on mxd/backend-service/src/main/java/org/eclipse/mxd/Main.java:
```diff
@@ -0,0 +1,74 @@
+    }
+
+    private static void startJettyServer() {
+        Properties appProperties = new Properties();
```
Create a `SettingResolver` class with `initialize()` and `getSetting(String key)` methods.
Inside initialize() load all properties into a `Map<String, String>` then load all `System.getEnv()` into the same map. (envs will have preference, so it will override properties).

ravinderkumarsap replied:
SettingResolver  is created with initialize() and getSetting(String key) methods is completed.

hemantxpatel commented on mxd/backend-service/src/main/java/org/eclipse/mxd/entity/Transfer.java:
```diff
@@ -0,0 +1,128 @@
+@Entity
+public class Transfer {
+
+    @Id
```
Primary key will be `transfer id`, no need of auto generated seq key.

ravinderkumarsap replied:
As suggested transferId is a primary key and auto generated seq key is removed

hemantxpatel commented on mxd/backend-service/src/main/java/org/eclipse/mxd/model/TransferRequest.java:
```diff
@@ -0,0 +1,94 @@
+    private String authKey;
+    private String authCode;
+    @JsonIgnore
+    private PropertiesModel properties;
```
Replace it with `Map<String, Object>`

ravinderkumarsap replied:
completed replace with Map as suggested

hemantxpatel commented on mxd/backend-service/src/main/java/org/eclipse/mxd/repository/Impl/TransferRepositoryImpl.java:
```diff
@@ -0,0 +1,145 @@
+        try {
+            Session session = HibernateUtil.getSessionFactory().openSession();
+            if (session != null) {
+                transfer = session.createQuery("SELECT t FROM Transfer t WHERE t.transferID = :transferID", Transfer.class)
```
Use transferId as primary key.

ravinderkumarsap replied:
completed terraform apply

hemantxpatel commented on mxd/backend-service/src/main/java/org/eclipse/mxd/service/Impl/ContentServiceImpl.java:
```diff
@@ -0,0 +1,165 @@
+            if (contentResponse != null && !contentResponse.isEmpty()) {
+                List<ContentModelResponse> contentsModels = new ArrayList<ContentModelResponse>();
+                for (ContentsRequest contentsRequest : contentResponse) {
+                    JsonNode jsonNode = new ObjectMapper().readTree(contentsRequest.getAsset());
```
Create `ObjectMapper` once and resuse it. Or create a Singleton class to return a same instance everywhere.

ravinderkumarsap replied:
used object mapper

hemantxpatel commented on mxd/backend-service/src/main/java/org/eclipse/mxd/model/PropertiesModel.java:
```diff
@@ -0,0 +1,28 @@
+ * 
+ ******************************************************************************/
+
+package org.eclipse.mxd.model;
```
This class is not needed at all.

ravinderkumarsap replied:
remove the unused class

hemantxpatel commented on mxd/backend-service/src/main/java/org/eclipse/mxd/service/Impl/ContentRandomServiceImpl.java:
```diff
@@ -0,0 +1,50 @@
+    @Override
+    public Response getRandomContent() {
+        try {
+            return contentRamdomRepository.getRandomContent();
```
`contentRamdomRepository` is not neeed, `repository` denotes a database access object. You can create random content here and return it.

ravinderkumarsap replied:
contentRamdomRepository is removed

hemantxpatel commented on mxd/backend-service/src/main/java/org/eclipse/mxd/service/Impl/ContentServiceImpl.java:
```diff
@@ -0,0 +1,165 @@
+        try {
+            ContentsRequest content = this.contentServiceRepository.getAssetById(id);
+            if (content != null && content.getId() != null) {
+                ContentJsonResponse contentModelRes = new ContentJsonResponse();
```
`ContentJsonResponse` is not needed, directly parse into a JsonNode and use it.

ravinderkumarsap replied:
passed the Json Node

hemantxpatel commented on mxd/backend-service/src/main/java/org/eclipse/mxd/service/Impl/TransferServiceImpl.java:
```diff
@@ -0,0 +1,154 @@
+            if (transfersModel != null && transfersModel.getId() != null) {
+                JsonNode jsonNode = new ObjectMapper().readTree(transfersModel.getAsset());
+                TransferModelsResponse transferModelsResponse = new TransferModelsResponse();
+                transferModelsResponse.setAsset(jsonNode);
```
There's no need of `TransferModelsResponse` class. You can directly use jsonNode.

ravinderkumarsap replied:
Now jsonNode is used directly.

hemantxpatel commented on mxd/backend-service/src/main/java/org/eclipse/mxd/service/Impl/TransferServiceImpl.java:
```diff
@@ -0,0 +1,154 @@
+            TransfersModel transfersModel = this.transferRepository.getTransferById(id);
+            if (transfersModel != null && transfersModel.getContents() != null) {
+                JsonNode jsonNode = new ObjectMapper().readTree(transfersModel.getContents());
+                TransferModelsResponse transferModelsResponse = new TransferModelsResponse();
```
No need of `TransferModelsResponse`, use jsonNode directly.

ravinderkumarsap replied:
I have used  jsonNode directly as suggested.

hemantxpatel commented on mxd/backend-service/src/main/java/org/eclipse/mxd/util/DateTimeUtil.java:
```diff
@@ -0,0 +1,36 @@
+
+public class DateTimeUtil {
+
+    public static Date getCurrentDateTimeInAsiaKolkata() {
```
Date class uses system timezone, so if server is in some other location, time will be in that zone. Hence method name is not suitable. You can use `new Date()` directly instead of this method. Also `Date` class has lots of issue. Check, if you could use Instant class.

ravinderkumarsap replied:
System time zone is removed and directly used new Date()

hemantxpatel commented on mxd/backend-service/src/main/java/org/eclipse/mxd/util/HibernateUtil.java:
```diff
@@ -0,0 +1,96 @@
+                Configuration configuration = new Configuration();
+                Properties settings = new Properties();
+                Properties appProperties = new Properties();
+                try (InputStream input = HibernateUtil.class.getClassLoader().getResourceAsStream("application.properties")) {
```
Don't load properties again, as earlier suggested, create a separate `SettingResolver` class.

ravinderkumarsap replied:
SettingResolver  class is created separately.

hemantxpatel commented on mxd/backend-service/src/main/java/org/eclipse/mxd/util/HibernateUtil.java:
```diff
@@ -0,0 +1,96 @@
+                logger.info(" env databaseUrl : " + databaseUrl);
+                logger.info(" env databaseUser : " + databaseUser);
+                logger.info(" env databasePassword : " + databasePassword);
+                logger.info(" props databaseUrl : " + appProperties.getProperty("database.url"));
```
Property key and env key should be same. Env value should take preference over property value.

ravinderkumarsap replied:
Now I have created SettingResolver class. In SettingResolver and override the properties with Env Variables

hemantxpatel commented on mxd/backend-service/src/main/java/org/eclipse/mxd/util/SettingResolver.java:
```diff
@@ -0,0 +1,91 @@
+            Properties appProperties = new Properties();
+            appProperties.load(input);
+
+            for (String key : appProperties.stringPropertyNames()) {
+                if(key.equalsIgnoreCase("database.url")){
+                    if(System.getenv("BACKEND-SERVICE-DATABASE_URL") != null){
+                        settings.put(key, System.getenv("BACKEND-SERVICE-DATABASE_URL") );
+                    }else{
+                        settings.put(key, appProperties.getProperty(key));
+                    }
+                }
+
+                if(key.equalsIgnoreCase("database.user")){
+                    if(System.getenv("BACKEND-SERVICE-DATABASE_USER") != null){
+                        settings.put(key, System.getenv("BACKEND-SERVICE-DATABASE_USER") );
+                    }else{
+                        settings.put(key, appProperties.getProperty(key));
+                    }
+                }
+
+                if(key.equalsIgnoreCase("database.password")){
+                    if(System.getenv("BACKEND-SERVICE-DATABASE_PASSWORD") != null){
+                        settings.put(key, System.getenv("BACKEND-SERVICE-DATABASE_PASSWORD") );
+                    }else{
+                        settings.put(key, appProperties.getProperty(key));
+                    }
```
There's no need for this.
```
Properties appProperties = new Properties();
appProperties.load(input);
settings.putAll(appProperties);
settings.putAll(System.getEnv());
```

Define env variables key same as property key i.e. `server.port`, `database.username` etc.

ravinderkumarsap replied:
Env variables are defined save as property key

hemantxpatel commented on mxd/backend-service/src/main/java/org/eclipse/mxd/util/SingletonObjectMapper.java:
```diff
@@ -0,0 +1,46 @@
+    }
+
+    public static ObjectMapper getObjectMapper() {
+        if (objectMapper == null) {
+                if (objectMapper == null) {
+                    objectMapper = SingletonObjectMapper.getObjectMapper();
+                }
+        }
```
Remove these lines, objectMapper already initialized, directly return it.

ravinderkumarsap replied:
removed unwanted lines.

hemantxpatel commented on mxd/backend-service/src/main/java/org/eclipse/mxd/util/SingletonObjectMapper.java:
```diff
@@ -0,0 +1,46 @@
+
+public class SingletonObjectMapper {
+
+    private static ObjectMapper objectMapper = new ObjectMapper();
```
Make it final

ravinderkumarsap replied:
it is marked final now.

hemantxpatel commented on mxd/backend-service.tf:
```diff
@@ -0,0 +1,88 @@
+            name           = "backend-port"
+          }
+          env {
+            name  = "BACKEND-SERVICE-DATABASE_URL"
```
Define env keys same as properties keys:
```
server.port=8080
database.driver=org.postgresql.Driver
database.url=jdbc:postgresql://localhost:5432/backendservicedb
database.user=postgres
database.password=postgres
```

ravinderkumarsap replied:
env keys and property key defined same now

ravinderkumarsap commented on mxd/backend-service/src/main/java/org/eclipse/mxd/Main.java:
```diff
@@ -0,0 +1,77 @@
+            appProperties.load(input);
+        } catch (Exception e) {
+            logger.info(" error " + e.getMessage());
+        }
```
removed unused code

ravinderkumarsap commented on mxd/backend-service/src/main/java/org/eclipse/mxd/Main.java:
```diff
@@ -0,0 +1,77 @@
+            logger.info(" error " + e.getMessage());
+        }
+        SettingResolver settingResolver = new SettingResolver();
+        settingResolver.initialize();
```
call initialize from constructor is completed

ravinderkumarsap commented on mxd/backend-service/src/main/java/org/eclipse/mxd/Main.java:
```diff
@@ -0,0 +1,77 @@
+        } catch (Exception e) {
+            logger.info(" error " + e.getMessage());
+        }
+        SettingResolver settingResolver = new SettingResolver();
```
SettingResolver class is marked singleton now

ravinderkumarsap commented on mxd/backend-service/src/main/java/org/eclipse/mxd/model/TransfersRequest.java:
```diff
@@ -0,0 +1,101 @@
+    private String endpoint;
+    private String authKey;
+    private String authCode;
+    @JsonIgnore
```
@JsonIgnore is removed

ravinderkumarsap commented on mxd/backend-service/src/main/java/org/eclipse/mxd/util/SettingResolver.java:
```diff
@@ -0,0 +1,91 @@
+public class SettingResolver {
+    private Map<String, String> settings;
+
+    public void initialize() {
```
initialize method is called from constructor and marked private.

paullatzelsperger commented on mxd/backend-service/src/main/java/org/eclipse/mxd/entity/Content.java:
```diff
@@ -0,0 +1,95 @@
+
+/*******************************************************************************
+ * Copyright (c) 2023 Bayerische Motoren Werke Aktiengesellschaft (BMW AG)
```
Is this Copyright correct?

ravinderkumarsap replied:
I have took Copyright from eclipse-tractusx and changed in all files now.

paullatzelsperger commented on mxd/backend-service/src/main/java/org/eclipse/mxd/entity/Content.java:
```diff
@@ -0,0 +1,95 @@
+ ******************************************************************************/
+
+
+package org.eclipse.mxd.entity;
```
package is wrong. must start with `org.eclipse.tractusx.`

ravinderkumarsap replied:
I have changed package org.eclipse.mxd to org.eclipse.tractusx as suggested

paullatzelsperger replied:
nope, it should _start_ with `org.eclipse.tractusx.`. Meaning, the correct package decl would then be e.g. `org.eclipse.tractusx.mxd.entity` etc.

paullatzelsperger reviewed (changes requested):
Package declarations are wrong, and many copyright headers are likely wrong. Further, there is no reason to introduce J2EE when EDC modules are perfectly capable of implementing a REST API.

ravinderkumarsap commented:
> Package declarations are wrong, and many copyright headers are likely wrong. Further, there is no reason to introduce J2EE when EDC modules are perfectly capable of implementing a REST API.



> Package declarations are wrong, and many copyright headers are likely wrong. Further, there is no reason to introduce J2EE when EDC modules are perfectly capable of implementing a REST API.

I have updated the package declarations as discussed and updated the copyright headers as suggested.

ravinderkumarsap commented on mxd/backend-service/src/main/java/org/eclipse/mxd/repository/Impl/ContentsRepositoryImpl.java:
```diff
@@ -0,0 +1,154 @@
+    }
+
+    public Content saveContent(String asset) {
+        logger.info("line no 62 saveContent : "+asset);
```
Removed the unused message.

ravinderkumarsap commented on mxd/backend-service/src/main/java/org/eclipse/mxd/repository/Impl/ContentsRepositoryImpl.java:
```diff
@@ -0,0 +1,154 @@
+                session.getTransaction().commit();
+                session.close();
+
+                logger.info("line no 76 saved content : "+content.toString());
```
added useful message.

ravinderkumarsap commented on mxd/backend-service/src/main/java/org/eclipse/mxd/repository/Impl/TransferRepositoryImpl.java:
```diff
@@ -0,0 +1,145 @@
+            Session session = HibernateUtil.getSessionFactory().openSession();
+            if (session != null) {
+                transfer = session.find(Transfer.class, id);
+              logger.info("line 83 transferRepoImpl "+transfer.toString());
```
added useful message

ravinderkumarsap commented on mxd/backend-service/src/main/java/org/eclipse/mxd/service/Impl/ContentServiceImpl.java:
```diff
@@ -0,0 +1,165 @@
+            Long id = contentServiceRepository.createAsset(requestBody);
+            logger.info("line no 132 contentId : "+id);
+            if (id != null) {
+                logger.info("line no 134 "+uriInfo.getAbsolutePath() + "");
```
added useful message

ravinderkumarsap commented on mxd/backend-service/src/main/java/org/eclipse/mxd/service/Impl/TransferServiceImpl.java:
```diff
@@ -0,0 +1,151 @@
+            TransfersResponse transfersResponse = this.transferRepository.getTransferById(id);
+            logger.info("line 89 transferImpl : "+transfersResponse);
+            if (transfersResponse != null && transfersResponse.getTransferID() != null) {
+                logger.info("line 91 transferImpl : "+transfersResponse.getAsset());
```
added useful message

ravinderkumarsap commented on mxd/backend-service/src/main/java/org/eclipse/mxd/util/SettingResolver.java:
```diff
@@ -0,0 +1,91 @@
+import java.util.Properties;
+
+public class SettingResolver {
+    private Map<String, String> settings;
```
i have made it static

ravinderkumarsap commented on mxd/backend-service/src/main/java/org/eclipse/mxd/repository/Impl/ContentsRepositoryImpl.java:
```diff
@@ -0,0 +1,154 @@
+        Content content = new Content();
+        try {
+            Session session = HibernateUtil.getSessionFactory().openSession();
+            logger.info("line no 66 saveContent get Session ");
```
added better message .

ravinderkumarsap commented on mxd/backend-service/src/main/java/org/eclipse/mxd/service/Impl/TransferServiceImpl.java:
```diff
@@ -0,0 +1,151 @@
+    public Response getTransfer(String id) {
+        try {
+            TransfersResponse transfersResponse = this.transferRepository.getTransferById(id);
+            logger.info("line 89 transferImpl : "+transfersResponse);
```
added meaning full message logger.

ravinderkumarsap commented on mxd/backend-service/src/main/java/org/eclipse/mxd/service/Impl/TransferServiceImpl.java:
```diff
@@ -0,0 +1,151 @@
+            if (transfersResponse != null && transfersResponse.getTransferID() != null) {
+                logger.info("line 91 transferImpl : "+transfersResponse.getAsset());
+                JsonNode jsonNode = SingletonObjectMapper.getObjectMapper().readTree(transfersResponse.getAsset());
+                logger.info("line 93 transferImpl : "+jsonNode.toString());
```
added meaning full message logger.

paullatzelsperger commented:
> > Package declarations are wrong, and many copyright headers are likely wrong. Further, there is no reason to introduce J2EE when EDC modules are perfectly capable of implementing a REST API.
> 
> > Package declarations are wrong, and many copyright headers are likely wrong. Further, there is no reason to introduce J2EE when EDC modules are perfectly capable of implementing a REST API.
> 
> I have updated the package declarations as discussed and updated the copyright headers as suggested.

you are still using Jakarta EE, for which there is no reason and which I asked you to replace earlier. Please use EDC modules and components, a lightweight REST API can easily be built with it, checkout [this](https://github.com/eclipse-edc/Connector/tree/68a5fd5349efffb8bf80580512f81c445edce207/system-tests/e2e-transfer-test/backend-service).

ravinderkumarsap commented:
> > > Package declarations are wrong, and many copyright headers are likely wrong. Further, there is no reason to introduce J2EE when EDC modules are perfectly capable of implementing a REST API.
> > 
> > 
> > > Package declarations are wrong, and many copyright headers are likely wrong. Further, there is no reason to introduce J2EE when EDC modules are perfectly capable of implementing a REST API.
> > 
> > 
> > I have updated the package declarations as discussed and updated the copyright headers as suggested.
> 
> you are still using Jakarta EE, for which there is no reason and which I asked you to replace earlier. Please use EDC modules and components, a lightweight REST API can easily be built with it, checkout [this](https://github.com/eclipse-edc/Connector/tree/68a5fd5349efffb8bf80580512f81c445edce207/system-tests/e2e-transfer-test/backend-service).

We have started with converting to EDC modules. We will update you once completed .

hemantxpatel commented on mxd/backend-service/Dockerfile:
```diff
@@ -0,0 +1,39 @@
+
+# Use the eclipse-temurin:17.0.8.1_1-jre-alpine as the base image
+FROM eclipse-temurin:17.0.8.1_1-jre-alpine
+ARG JAR
```
Remove this unused Arg `JAR`

ravinderkumarsap replied:
unused Arg JAR is removed

hemantxpatel commented on mxd/backend-service.tf:
```diff
@@ -0,0 +1,85 @@
+      spec {
+        container {
+          name              = "backend-service"
+          image             = "tractusx/mxd-backend-service:1.0.0"
```
Rename image file. Docker image getting created as `tractusx/backend-service`

ravinderkumarsap replied:
Docker image file renamed as mentioned.

hemantxpatel commented on mxd/backend-service.tf:
```diff
@@ -0,0 +1,85 @@
+            value = "postgres"
+          }
+          env {
+            name  = "edc.datasource.backendservice..password"
```
There's an extra dot in env name.

ravinderkumarsap replied:
extra dot is removed.

hemantxpatel commented on mxd/backend-service/Dockerfile:
```diff
@@ -0,0 +1,39 @@
+# application.properties
```
Remove line `# application.properties`

ravinderkumarsap replied:
# application.properties line is removed

hemantxpatel commented on mxd/backend-service/build.gradle.kts:
```diff
@@ -0,0 +1,91 @@
+}
+
+val dockerTask: DockerBuildImage = tasks.create("dockerize", DockerBuildImage::class) {
+    buildArgs.put("JAR", "${project.buildDir}/libs/${project.name}.jar")
```
Remove this unused arg `JAR`

ravinderkumarsap replied:
unused arg JAR is removed

hemantxpatel commented on mxd/backend-service/build.gradle.kts:
```diff
@@ -0,0 +1,91 @@
+
+val dockerTask: DockerBuildImage = tasks.create("dockerize", DockerBuildImage::class) {
+    buildArgs.put("JAR", "${project.buildDir}/libs/${project.name}.jar")
+    println("build 1 : ${buildArgs.get()}")
```
Remove println

ravinderkumarsap replied:
println is removed

hemantxpatel commented on mxd/backend-service/src/main/java/org/eclipse/tractusx/mxd/backendservice/controller/ContentApiController.java:
```diff
@@ -0,0 +1,92 @@
+
+
+    private String createJsonResponse(String id, UriInfo uriInfo) {
+        ObjectMapper objectMapper = new ObjectMapper();
```
Create a shared ObjectMapper or create one in Extension and inject it via constructor.

ravinderkumarsap replied:
I have injected ObjectMapper via constructor.

hemantxpatel commented on mxd/backend-service/src/main/java/org/eclipse/tractusx/mxd/backendservice/controller/ContentApiController.java:
```diff
@@ -0,0 +1,92 @@
+        ObjectMapper objectMapper = new ObjectMapper();
+        JsonNode jsonResponse = objectMapper.createObjectNode()
+                .put("id", id)
+                .put("url", uriInfo.getBaseUri() + "v1/contents/" + id);
```
Request url may or may not have a trailing slash `/`. Handle this.

ravinderkumarsap replied:
I have handled Request Url slash using UrlBuilder.

hemantxpatel commented on mxd/backend-service/src/main/java/org/eclipse/tractusx/mxd/backendservice/service/HttpConnectionService.java:
```diff
@@ -0,0 +1,62 @@
+            }
+
+        } catch (Exception e) {
+
```
Add some log here. Log exception

ravinderkumarsap replied:
I have used monitor to log the exception.

ravinderkumarsap commented:
> > > > Package declarations are wrong, and many copyright headers are likely wrong. Further, there is no reason to introduce J2EE when EDC modules are perfectly capable of implementing a REST API.
> > > 
> > > 
> > > > Package declarations are wrong, and many copyright headers are likely wrong. Further, there is no reason to introduce J2EE when EDC modules are perfectly capable of implementing a REST API.
> > > 
> > > 
> > > I have updated the package declarations as discussed and updated the copyright headers as suggested.
> > 
> > 
> > you are still using Jakarta EE, for which there is no reason and which I asked you to replace earlier. Please use EDC modules and components, a lightweight REST API can easily be built with it, checkout [this](https://github.com/eclipse-edc/Connector/tree/68a5fd5349efffb8bf80580512f81c445edce207/system-tests/e2e-transfer-test/backend-service).
> 
> We have started with converting to EDC modules. We will update you once completed.

I have completed the backend service using EDC modules. Let me know the feedback.

ravinderkumarsap commented:
@paullatzelsperger   Could you please review and provide your feedback.

ndr-brt reviewed (changes requested):
I started adding some comments, but it looks like there are some lacks about the EDC extension model, and, more importantly, automated tests are missing, please add at least some E2E tests that cover all the flow, from creating a resource (content/transfer) and fetching them, to see that everything's working as expected. without them there's no point in going down to a proper review

stephanbcbauer commented:
@ravinderkumarsap what about this PR? Do you still work on this? thx

ravinderkumarsap commented:
> @ravinderkumarsap what about this PR? Do you still work on this? thx

@stephanbcbauer  yes. I have received some review comments yesterday. I am working on review comments.

ravinderkumarsap commented on .github/workflows/run-terraform.yml:
```diff
@@ -53,7 +53,15 @@ jobs:
+      - name: Setup JDK 17
+        uses: actions/setup-java@v3.13.0
+        with:
+          java-version: '17'
```
I have updated the java version 21 for run-terraform file.

ravinderkumarsap commented on mxd/backend-service/Dockerfile:
```diff
@@ -0,0 +1,37 @@
+# ******************************************************************************/
+
+# Use the eclipse-temurin:17.0.8.1_1-jre-alpine as the base image
+FROM eclipse-temurin:17.0.8.1_1-jre-alpine
```
Java version 21 has been updated for Docker files.

ravinderkumarsap commented on mxd/backend-service/build.gradle.kts:
```diff
@@ -0,0 +1,89 @@
+    implementation("org.eclipse.edc:sql-core:0.4.1")
+    implementation("org.apache.commons:commons-dbcp2:2.11.0")
+    implementation("org.postgresql:postgresql:42.7.1")
+    implementation("org.eclipse.edc:transform-core:0.4.1")
```
I have used the libraries file (libs.versions.toml) from Gradle and added the catalog and version and used in the build.gradle.kts

ravinderkumarsap commented on mxd/backend-service/src/main/java/org/eclipse/tractusx/mxd/backendservice/service/HttpConnectionService.java:
```diff
@@ -0,0 +1,68 @@
+    public String getUrlAssets(Transfer receivedModel,Monitor monitor) {
+        String res = null;
+        monitor.info("getUrlAssets "+receivedModel);
+        OkHttpClient client = new OkHttpClient();
```
I have used the already initialized OkHttpClient for for httpclient calls

ravinderkumarsap commented on mxd/backend-service/src/main/java/org/eclipse/tractusx/mxd/backendservice/BackendServiceExtension.java:
```diff
@@ -0,0 +1,208 @@
+    private TransactionContext transactionContext;
+    @Inject
+    private TypeManager typeManager;
+    @Inject(required = false)
+    private ContentStatementsService statements;
+
+    @Inject(required = false)
+    private TransferStatementsService transferStatementsService;
```
I am using this ContentStatementsService  and TransferStatementsService  for saving the content and transfer to DB

ravinderkumarsap commented on mxd/backend-service/Dockerfile:
```diff
@@ -0,0 +1,37 @@
+#*******************************************************************************
+# * Copyright (c) 2023 Bayerische Motoren Werke Aktiengesellschaft (BMW AG)
```
I have updated all the files with new year as 2024.

ravinderkumarsap commented on mxd/backend-service/gradle/wrapper/gradle-wrapper.properties:
```diff
@@ -0,0 +1,6 @@
+#Mon Dec 18 15:03:56 IST 2023
+distributionBase=GRADLE_USER_HOME
+distributionPath=wrapper/dists
+distributionUrl=https\://services.gradle.org/distributions/gradle-8.2-bin.zip
```
I have updated the gradle distributions version and using  to latest version 8.6 now.

ravinderkumarsap commented on mxd/backend-service/src/main/java/org/eclipse/tractusx/mxd/util/Converter.java:
```diff
@@ -0,0 +1,39 @@
+            return null;
+        }
+        try {
+            return object instanceof String ? object.toString() : new ObjectMapper().writeValueAsString(object);
```
I am using the already initialized ObjectMapper now in place of new ObjectMapper object every time.

ravinderkumarsap commented on mxd/backend-service/build.gradle.kts:
```diff
@@ -0,0 +1,89 @@
+    testImplementation("org.junit.jupiter:junit-jupiter")
+    implementation(libs.edc.configuration.filesystem)
+    implementation("org.eclipse.edc:boot:0.4.1")
+    implementation("org.eclipse.edc:connector-core:0.4.1")
```
I am using the httpclient from the connector-core. So that i have using it

ravinderkumarsap commented on mxd/backend-service/src/main/java/org/eclipse/tractusx/mxd/backendservice/BackendServiceExtension.java:
```diff
@@ -0,0 +1,208 @@
+                    getTransferStatementImpl(),
+                    context.getMonitor());
+            context.registerService(ContentStoreService.class, sqlContentStore);
+            context.registerService(TypeTransformerRegistry.class, transformerRegistry);
```
I have removed the transformer Registry.

ravinderkumarsap commented:
@ndr-brt  I have resolved all the review comment . Let me know in case of anything need to cover. I will cover it.

ndr-brt commented:
@ravinderkumarsap as I said before, please add E2E test coverage that demonstrates that the integration is working, after that I'll review the PR properly

ravinderkumarsap commented:
> @ravinderkumarsap as I said before, please add E2E test coverage that demonstrates that the integration is working, after that I'll review the PR properly.

@ndr-brt I have completed the review comments and updated on PR. I am working on the E2E test coverage.

ndr-brt commented:
> > @ravinderkumarsap as I said before, please add E2E test coverage that demonstrates that the integration is working, after that I'll review the PR properly.
> 
> @ndr-brt I have completed the review comments and updated on PR. I am working on the E2E test coverage.

when the tests will be ready please re-request my review

ndr-brt reviewed (changes requested):
> > @ravinderkumarsap as I said before, please add E2E test coverage that demonstrates that the integration is working, after that I'll review the PR properly.
> 
> @ndr-brt I have completed the review comments and updated on PR. I am working on the E2E test coverage.

when the tests will be ready please re-request my review

ndr-brt reviewed (changes requested):
please ensure that tests are running as part of the CI

ieuna commented on mxd/backend-service/src/main/java/org/eclipse/tractusx/mxd/backendservice/BackendServiceExtension.java:
```diff
@@ -0,0 +1,256 @@
+                .logOnFailedAttempt(true)
+                .build();
+        return (RetryPolicy<T>) RetryPolicyFactory.create(configuration, context.getMonitor());
+    }
```
Removed custom configurations for these services.

ieuna commented on mxd/backend-service/src/test/java/org/eclipse/tractusx/mxd/e2e/ContentApiEndToEndTest.java:
```diff
@@ -0,0 +1,152 @@
+public class ContentApiEndToEndTest {
+
+    @Nested
+    @PostgresqlIntegrationTest
```
Tests are passing now. These are also part of workflow declared in verify-backend-service.yaml now.

ieuna commented on mxd/backend-service/build.gradle.kts:
```diff
@@ -0,0 +1,111 @@
+}
+
+dependencies {
+    implementation(libs.restAssured)
+    implementation(libs.edc.configuration.filesystem)
+    implementation(libs.edc.boot)
+    implementation(libs.edc.json.ld)
+    implementation(libs.edc.web.spi)
+    implementation(libs.edc.api.core)
+    implementation(libs.edc.core)
+    implementation(libs.edc.http)
+    implementation(libs.edc.http.lib)
+    implementation(libs.edc.http.spi)
+    implementation(libs.edc.jersey.core)
+    implementation(libs.swagger.core)
+    implementation(libs.edc.transaction.spi)
+    implementation(libs.edc.connector.core)
+    implementation(libs.edc.sql.core)
+    implementation(libs.apache.commons)
+    implementation(libs.postgresql)
+    implementation(libs.edc.transform)
+    implementation(libs.edc.transaction)
+    implementation(libs.edc.util)
```
Removed unused libraries

ieuna commented on mxd/backend-service/gradle/libs.versions.toml:
```diff
@@ -0,0 +1,48 @@
+format.version = "1.1"
+[versions]
+assertj = "3.25.3"
+edc = "0.6.0"
```
Updated EDC version to 0.7.0

ieuna commented on mxd/backend-service/build.gradle.kts:
```diff
@@ -0,0 +1,111 @@
+    testImplementation(libs.assertj)
+    testImplementation(libs.edc.junit)
+    implementation(libs.edc.junit.base)
+    testImplementation(libs.testng)
```
Removed unused libraries.

ieuna commented on mxd/backend-service/build.gradle.kts:
```diff
@@ -0,0 +1,111 @@
+    testImplementation(libs.postgres.containers)
+
+    testImplementation(testFixtures(libs.edc.sql.core))
+    testImplementation(libs.edc.core)
```
Removed unused libraries.

ieuna commented on mxd/backend-service/src/main/java/org/eclipse/tractusx/mxd/backendservice/BackendServiceExtension.java:
```diff
@@ -0,0 +1,226 @@
+
+    public static final String NAME = "Backend Services";
+
+    @Inject(required = false)
+    private EventListener okHttpEventListener;
```
Removed.

ieuna commented on mxd/backend-service/src/main/java/org/eclipse/tractusx/mxd/backendservice/BackendServiceExtension.java:
```diff
@@ -0,0 +1,226 @@
+    @Inject
+    private TypeManager typeManager;
+
+    @Inject
+    private TypeTransformerRegistry transformerRegistry;
```
Removed.

ieuna commented on mxd/backend-service/src/main/java/org/eclipse/tractusx/mxd/backendservice/BackendServiceExtension.java:
```diff
@@ -0,0 +1,226 @@
+                    getStatementImpl(),
+                    monitor);
+            //register content service
+            context.registerService(ContentStoreService.class, sqlContentStore);
```
Needed by test classes.

ndr-brt replied:
@ieuna if so there's the need to change the test approach, production code should not do things that are only needed by tests, but it should be effectively testable.

e.g. currently tests are getting these services to save new objects in the store, but they can do that by calling the `POST` method on the API, that would remove unnecessary service registration

ieuna replied:
Removed service registration

ieuna commented on mxd/backend-service/src/main/java/org/eclipse/tractusx/mxd/backendservice/BackendServiceExtension.java:
```diff
@@ -0,0 +1,226 @@
+            //register content service
+            context.registerService(ContentStoreService.class, sqlContentStore);
+            ContentService contentService = new ContentServiceImpl(sqlContentStore, monitor);
+            context.registerService(ContentService.class, contentService);
```
Needed by test classes.

ieuna commented on mxd/backend-service/src/main/java/org/eclipse/tractusx/mxd/backendservice/BackendServiceExtension.java:
```diff
@@ -0,0 +1,226 @@
+                    monitor);
+
+            //register transfer service
+            context.registerService(TransferStoreService.class, sqlTransferStore);
```
Needed by test classes.

ieuna commented on mxd/backend-service/src/main/java/org/eclipse/tractusx/mxd/backendservice/BackendServiceExtension.java:
```diff
@@ -0,0 +1,226 @@
+                    edcHttpClient,
+                    monitor,
+                    getHttConnectionService(edcHttpClient, monitor));
+            context.registerService(TransferService.class, transferService);
```
Removed.

ieuna commented on mxd/backend-service/src/main/java/org/eclipse/tractusx/mxd/backendservice/BackendServiceExtension.java:
```diff
@@ -0,0 +1,226 @@
+        return new ContentStatementsServiceImpl();
+    }
+
+    private TransferStatementsService getTransferStatementImpl() {
```
Removed one liners.

ieuna commented on mxd/backend-service/src/main/java/org/eclipse/tractusx/mxd/backendservice/BackendServiceExtension.java:
```diff
@@ -0,0 +1,226 @@
+        dataSourceRegistry.register(Constants.DATA_SOURCE_NAME, createLocalDataSource(context));
+    }
+
+    @Provider
+    public EdcHttpClient createEdcHttpClient(ServiceExtensionContext context) {
+        return new EdcHttpClientImpl(new OkHttpClient(), RetryPolicy.ofDefaults(), context.getMonitor());
+    }
```
Removed, EdcHttpClient is being injected now.

ieuna commented on mxd/backend-service/src/main/java/org/eclipse/tractusx/mxd/backendservice/BackendServiceExtension.java:
```diff
@@ -0,0 +1,226 @@
+        return new EdcHttpClientImpl(new OkHttpClient(), RetryPolicy.ofDefaults(), context.getMonitor());
+    }
+
+    public int runQuery(String query, Connection connection) {
+        return transactionContext.execute(() -> queryExecutor.execute(connection, query));
+    }
```
Removed.

ieuna commented on mxd/backend-service/src/main/java/org/eclipse/tractusx/mxd/backendservice/BackendServiceExtension.java:
```diff
@@ -0,0 +1,226 @@
+        return objectMapper;
+    }
+
+    public HttpConnectionService getHttConnectionService(EdcHttpClient httpClient, Monitor monitor) {
```
Removed.

ieuna commented on mxd/backend-service/src/main/java/org/eclipse/tractusx/mxd/backendservice/controller/ContentApiController.java:
```diff
@@ -0,0 +1,95 @@
+        try {
+            return objectMapper.writeValueAsString(jsonResponse);
+        } catch (IOException e) {
+            return "{\"error\": \"" + e.getMessage() + "\"}";
```
Added throw statement for (wrapped) EdcException in catch block.

ieuna commented on mxd/backend-service/src/main/java/org/eclipse/tractusx/mxd/backendservice/service/ContentService.java:
```diff
@@ -0,0 +1,30 @@
+
+    String getAllContent();
+
+    StoreResult<ContentResponse> getContent(String contentId);
```
Updated type to return ServiceResult.

ieuna commented on mxd/backend-service/src/main/java/org/eclipse/tractusx/mxd/backendservice/entity/Content.java:
```diff
@@ -0,0 +1,71 @@
+            return this;
+        }
+
+        @Override
+        public Content build() {
+            super.build();
+
+            return entity;
+        }
```
Removed.

ieuna commented on mxd/backend-service/src/main/java/org/eclipse/tractusx/mxd/backendservice/entity/Transfer.java:
```diff
@@ -0,0 +1,104 @@
+            return this;
+        }
+
+        @Override
+        public Transfer build() {
+            super.build();
+            return entity;
+        }
```
Updated.

ieuna commented on mxd/backend-service/src/main/java/org/eclipse/tractusx/mxd/backendservice/service/ContentServiceImpl.java:
```diff
@@ -0,0 +1,58 @@
+
+    @Override
+    public String getAllContent() {
+        return contentStoreService.findAll(new QuerySpec()).map(ContentResponse::getData).toList().toString();
```
Updated to return list.

ieuna commented on mxd/backend-service/src/main/java/org/eclipse/tractusx/mxd/backendservice/service/HttpConnectionService.java:
```diff
@@ -0,0 +1,64 @@
+    }
+
+    public String getUrlAssets(Transfer receivedModel, Monitor monitor) {
+        var res = "";
```
Removed.

ieuna commented on mxd/backend-service/src/main/java/org/eclipse/tractusx/mxd/backendservice/service/TransferService.java:
```diff
@@ -0,0 +1,35 @@
+
+    List<TransferResponse> getAllTransfer();
+
+    StoreResult<TransferResponse> getTransfer(String transferId);
+
+    StoreResult<TransferResponse> getTransferContent(String transferId);
```
Updated type to return ServiceResult.

ieuna commented on mxd/backend-service/src/main/java/org/eclipse/tractusx/mxd/backendservice/statements/ContentStatementsServiceImpl.java:
```diff
@@ -0,0 +1,74 @@
+    @Override
+    public SqlQueryStatement createQuery(QuerySpec querySpec) {
+        var select = format("SELECT * FROM %s", getContentTable());
+        //return new SqlQueryStatement(select, querySpec, null);
```
Removed commented codes.

ieuna commented on mxd/backend-service/src/main/java/org/eclipse/tractusx/mxd/backendservice/statements/TransferStatementsServiceImpl.java:
```diff
@@ -0,0 +1,71 @@
+    @Override
+    public SqlQueryStatement createQuery(QuerySpec querySpec) {
+        var select = format("SELECT * FROM %s", getTransferTable());
+        //return new SqlQueryStatement(select, querySpec, null);
```
Removed commented codes.

ieuna commented on mxd/backend-service/src/main/java/org/eclipse/tractusx/mxd/backendservice/store/SqlContentStoreServiceImpl.java:
```diff
@@ -0,0 +1,140 @@
+        return queryExecutor.single(connection, false, this::mapResultSet, sql, id);
+    }
+
+    @Override
+    public StoreResult<Void> update(Content definition) {
+        return null;
+    }
+
+    @Override
+    public StoreResult<Content> deleteById(String id) {
+        return null;
+    }
```
Removed unused codes.

ieuna commented on mxd/backend-service/src/main/java/org/eclipse/tractusx/mxd/backendservice/store/SqlContentStoreServiceImpl.java:
```diff
@@ -0,0 +1,140 @@
+            monitor.warning(exception.getMessage());
+            throw new EdcPersistenceException(exception.getMessage(), exception);
+        }
+        return null;
```
Updated to throw exception.

ieuna commented on mxd/backend-service/src/main/java/org/eclipse/tractusx/mxd/backendservice/store/SqlTransferStoreServiceImpl.java:
```diff
@@ -0,0 +1,132 @@
+        return null;
+    }
+
+    @Override
+    public StoreResult<Void> update(Transfer transfer) {
+        return null;
+    }
+
+    @Override
+    public StoreResult<Transfer> deleteById(String id) {
+        return null;
+    }
```
Removed unused codes.

ieuna commented on mxd/backend-service/src/test/java/org/eclipse/tractusx/mxd/testfixtures/PostgresRuntime.java:
```diff
@@ -0,0 +1,39 @@
+
+    @RegisterExtension
+    public static final EdcRuntimeExtension RUNTIME = new EdcRuntimeExtension(
+            "",
```
This parameter is needed by EdcRuntimeExtension because of constructor overloading.

ndr-brt replied:
@ieuna nope, there's a different constructor where the modules are passed as varargs as last argument, so you could use that one instead (way cleaner: an empty string is confusing to read)

ieuna replied:
not passing any argument to the varargs wasn't working, so "new String()" is passed to the third argument(varargs modules) now.

```
@RegisterExtension
EdcRuntimeExtension RUNTIME = new EdcRuntimeExtension(
        "backend",
        Map.of(....),
        new String()
);
```

ndr-brt replied:
@ieuna you don't need the third argument, just two: runtime name and configuration

ieuna replied:
Updated.

ndr-brt reviewed (approved):
LGTM

ndr-brt commented:
please fix the conflicts and I will merge this

ieuna commented:
> please fix the conflicts and I will merge this

resolved

ndr-brt commented:
@ieuna there's a validate-terraform-format check failing

ieuna commented:
> @ieuna there's a validate-terraform-format check failing

fixed terraform format. validate terraform check is passing now
